File: data-structure/c-cpp/c/SingleLinkedList/DummyTailSingleLinkedList/DummyTailSingleLinkedList.c
#include "DummyTailSingleLinkedList.h"
#include <stdlib.h>
void init(List* list) {
    Node*  dummy_node = (Node*)malloc(sizeof(Node));
    dummy_node -> prev = NULL;
    list -> tail = dummy_node;
    list -> numOfData = 0;
}
void insert(List* list, Data data) {
    Node* new_node = (Node*)malloc(sizeof(Node));
    new_node -> data = data;
    new_node -> prev = list -> tail -> prev;
    list -> tail -> prev = new_node;
    list -> numOfData++;
}
int first(List* list, Data* data) {
    if (list -> numOfData == 0) return FALSE;
    list -> before = NULL;
    list -> cur = list -> tail;
    *data = list -> cur -> data;
    return TRUE;
}
int next(List* list, Data* data) {
    if (list -> cur -> prev == NULL) return FALSE;
    list -> before = list -> cur;
    list -> cur = list -> cur -> prev;
    *data = list -> cur -> data;
    return TRUE;
}

Data remove(List* list) {
    Node* r_pos = list -> cur;
    Data r_data = r_pos -> data;
    list -> before -> prev = list -> cur -> prev;
    free(r_pos);
    list -> numOfData--;
    return r_data;
}

int count(List* list) {
    return list -> numOfData;
}

File: data-structure/java/src/linkedlist/singlelinkedlist/tailsinglelinkedlist/dummy/DummyTailSingleLinkedList.java
package linkedlist.singlelinkedlist.tailsinglelinkedlist.dummy;

import exception.NoElementException;
import linkedlist.singlelinkedlist.tailsinglelinkedlist.Node;

public class DummyTailSingleLinkedList <T>{
    private Node<T> tail;
    private Node<T> prev;
    private Node<T> cur;
    private int numOfElements;

    DummyTailSingleLinkedList(){
        this.tail = new Node<>(null);
        this.prev = null;
        this.cur = null;
        this.numOfElements = 0;
    }

    public void insert(T data){
        Node<T> newNode = new Node<>(data);
        newNode.before = this.tail;
        this.tail = newNode;
        this.numOfElements++;
    }

    public T first(){
        if(this.tail.before == null) throw new NoElementException();
        this.prev = this.tail;
        this.cur = this.tail.before;
        return this.cur.data;
    }
    public T next(){
        if(this.cur.before == null) throw new NoElementException();
        this.prev = this.cur;
        this.cur = this.cur.before;
        return this.cur.data;
    }
    public T remove(){
        Node<T> rPos = this.cur;
        T rData = rPos.data;
        this.prev.before = this.cur.before;
        this.numOfElements--;
        return rData;

    }

    public int count(){
        return this.numOfElements;
    }


}
